### ISSUE TITLE
Poor documentation \ overall tool state

### ISSUE THREAD

**Issue by AlexanderKormilin** [markabrahams/node-net-snmp#251]
Hi there!

Tested against 3 different devices, all with snmpv3 (unable to use other cersions), all of the attempts failed with useless error messages (either auth error on valid credentials or random pduEncryption error, which in turn seem to be just a code error.

Is it supposed to work and im just unlucky or it is not supposed to work ok with snmp v3?
The pduEncryption errors i see are from attempt to address non existent prop, 
var aes = Encryption.algorithms[privProtocol]

which sure fails. I see the repo was updated not so long ago, but i can not figure out why is it failing to work on such easy scenarios.

Bes regradsm Aleksandr

**Comment by mcm1957** [comment]
Don't know your problem in detail - I can just answer the questions whetehr its supposed to work :-) yes it does.

node lib works with ioBroker.snmp and snmpV3 i.e. against a synology dsm bod

**Comment by christopher-nielsen-ny** [comment]
> Don't know your problem in detail - I can just answer the questions whether its supposed to work :-) yes it does.
> 
> node lib works with ioBroker.snmp and snmpV3 i.e. against a synology dsm bod

We've been using `node-net-snmp` within NodeJS as well, and it works beautifully. (Though with the caveat that we're not using authentication, so can't speak to that area specifically.)

Regardless, @markabrahams has been really great to work with, and he's always willing to assist. So if you can provide a minimal test case - along with specifics regarding the device(s) you're trying to connect to - that would be a good next step.

**Comment by AlexanderKormilin** [comment]
Allright.
First  - sorry for being too rude and emotional, it seems like currently emotional stability might be my problem. Not an excuse, but an attempt to be honest.

You've responded in a constructive way to not very constructive comment - thank you for being that cool and sorry again.

Second: the details.
Ive wrote a very POC'ish and simple module with two methods:


`const exporter = module.exports = {};

const snmp = require ("net-snmp");

const addr = 'ipv4 address'

const config = {
    port: 161,
    retries: 1,
    timeout: 5000,
    version: snmp.Version3,
    backwardsGetNexts: true,
    reportOidMismatchErrors: false,
    idBitsSize: 32,
};

const user = {
    name: "username",
    level: snmp.SecurityLevel.authPriv,
    authProtocol: snmp.AuthProtocols['sha'],
    authKey: "authkey",
    privProtocol: snmp.PrivProtocols['aes'],
    privKey: "privkey"
};

exporter.createSession = async function createSession(userObj, configObj) {
    return snmp.createV3Session (addr, userObj, configObj);
};

exporter.getMetrics = async function getMetrics() {
    const session = await this.createSession(user, config);
    const oids = ["1.3.6.1.2.1.43.11.1.1.8.1.4"];
    session.get (oids, function (error, varbinds) {
        if (error) {
            console.error (error);
        } else {
            for (var i = 0; i < varbinds.length; i++) {
                if (snmp.isVarbindError (varbinds[i])) {
                    console.error (snmp.varbindError (varbinds[i]));
                } else {
                    console.log (varbinds[i].oid + " = " + varbinds[i].value);
                }
            }
        }
        session.close ();
    });
}
`

The invocation of the getMetrics method leads to the following error: 
TypeError: Cannot read properties of undefined (reading 'encryptPdu')
    at Encryption.encryptPdu (/Users/aleksandrkormilin/WebstormProjects/kyoceraprometheusexporter/node_modules/net-snmp/index.js:1038:75)
    at Message.toBufferV3 (/Users/aleksandrkormilin/WebstormProjects/kyoceraprometheusexporter/node_modules/net-snmp/index.js:1364:33)
    at Message.toBuffer (/Users/aleksandrkormilin/WebstormProjects/kyoceraprometheusexporter/node_modules/net-snmp/index.js:1313:15)
    at Session.send (/Users/aleksandrkormilin/WebstormProjects/kyoceraprometheusexporter/node_modules/net-snmp/index.js:2210:28)
    at Session.sendV3Req (/Users/aleksandrkormilin/WebstormProjects/kyoceraprometheusexporter/node_modules/net-snmp/index.js:2674:7)
    at Session.onMsg (/Users/aleksandrkormilin/WebstormProjects/kyoceraprometheusexporter/node_modules/net-snmp/index.js:2150:9)
    at Socket.emit (node:events:514:28)
    at UDP.onMessage [as onmessage] (node:dgram:933:8)
    at UDP.callbackTrampoline (node:internal/async_hooks:130:17)

Im totally sure the credentials are valid, the address is correct, and the authentication options are set correctly, as im totally able to get the metrics from the very that device with the very that creds and config via the Mib Browser application.

Basically, thats all ive got right now.
Would really appreciate any thoughts or directions.

Best regards, Aleksandr.

**Comment by AlexanderKormilin** [comment]
And yeah, im sure aware there is a snmp exporter ready for the prometheus, im using it currently

The idea behind making a friends out of node and snmp is to be able to calculate device's status with some custom logic. Sound weird, i know, i just want to get rid of all the complex metrics aggregation and calculation in prom\grafana, and be able to define device status with some custom logic.

**Comment by mcm1957** [comment]
Not sure if you array access is working. Does  snmp.AuthProtocols['sha'] retun a valid value? 

I use the following construct

```
        let snmpSecurityLevel = 0;
        if ( pCTX.authSecLvl == 1 ) {
            snmpSecurityLevel = snmp.SecurityLevel.noAuthNoPriv; // no message authentication or encryption
        } else if ( pCTX.authSecLvl == 2 ) {
            snmpSecurityLevel = snmp.SecurityLevel.authNoPriv;  // message authentication and no encryption
        } else if ( pCTX.authSecLvl == 3 ) {
            snmpSecurityLevel = snmp.SecurityLevel.authPriv;    //for message authentication and encryption
        }

        let snmpAuthProtocol = 0;
        switch (Number(pCTX.authAuthProto)) {   /* ensure numeric type */
            default:
                snmpAuthProtocol = 0;
                break;
            case MD5:
                snmpAuthProtocol = snmp.AuthProtocols.md5;
                break;
            case SHA:
                snmpAuthProtocol = snmp.AuthProtocols.sha;
                break;
            case SHA224:
                snmpAuthProtocol = snmp.AuthProtocols.sha224;
                break;
            case SHA256:
                snmpAuthProtocol = snmp.AuthProtocols.sha256;
                break;
            case SHA384:
                snmpAuthProtocol = snmp.AuthProtocols.sha384;
                break;
            case SHA512:
                snmpAuthProtocol = snmp.AuthProtocols.sha512;
                break;
        }

        let snmpPrivProtocol = 0;
        if ( pCTX.authEncProto == DES ) {
            snmpPrivProtocol = snmp.PrivProtocols.des; // DES encryption
        } else if ( pCTX.authEncProto == AES ) {
            snmpPrivProtocol = snmp.PrivProtocols.aes; // AES encryption
        } else if ( pCTX.authEncProto == AES256B ) {
            snmpPrivProtocol = snmp.PrivProtocols.aes256b; // AES encryption
        } else if ( pCTX.authEncProto == AES256R ) {
            snmpPrivProtocol = snmp.PrivProtocols.aes256r; // AES encryption
        }

        const snmpUser = {
            name: pCTX.authUser,
            level: snmpSecurityLevel,
            authProtocol: snmpAuthProtocol,
            authKey: pCTX.authAuthKey,
            privProtocol: snmpPrivProtocol,
            privKey: pCTX.authEncKey
        };

        const snmpTransport = pCTX.isIPv6 ? 'udp6' : 'udp4';
        const snmpVersion = snmp.Version3;
        // ??? engineID: "8000B98380XXXXXXXXXXXXXXXXXXXXXXXX", // where the X's are random hex digits

        ret.session = snmp.createV3Session(pCTX.ipAddr, snmpUser, {
            port: pCTX.ipPort,   // default:161
            retries: 1,
            timeout: pCTX.timeout,
            backoff: 1.0,
            transport: snmpTransport,
            //trapPort: 162,
            version: snmpVersion,
            backwardsGetNexts: true,
            idBitsSize: 32,
            context: ''
        });
    } else {
        adapter.log.error('unsupported snmp version code (' + pCTX.snmpVers + ') for device "' + pCTX.name + '" (' + pCTX.ip + ')');
    }
```

see https://github.com/iobroker-community-adapters/ioBroker.snmp/blob/bd8dc74d39368ddda56f8eef45731195b1bce1cf/main.js#L1234

**Comment by AlexanderKormilin** [comment]
and youve probably spotted, but the device im querying (or trying to) is Kyocera's printer

**Comment by AlexanderKormilin** [comment]
> Does snmp.AuthProtocols['sha'] retun a valid value?

It does:

<img width="1037" alt="Снимок экрана 2024-04-03 в 21 16 44" src="https://github.com/markabrahams/node-net-snmp/assets/63704929/42357eaf-bff0-4c99-91cf-62a3fc62e7e7">


> And you initialise an object nemed i.e. 'cont user ...' while passing userObj as parameter.

Sorry, not sure i got your point. What's that different in a ways we are declaring our user's object?

**Comment by AlexanderKormilin** [comment]
It might be i was just too sporadic trying to figure whats going on here

ive reinstalled the net-snmp module with npm and it just started working fine, without any code changes or updates, im able to fetch metrics now.

**Comment by AlexanderKormilin** [comment]
Sorry, it seems like its just me having a hard time

I will make sure everything works as expected now and close the issue if no more issues are seen.

**Comment by mcm1957** [comment]
>Sorry, not sure i got your point. What's that different in a ways we are declaring our user's object?

// declaring a const object named user
const **user** = {
    name: "username",0    level: snmp.SecurityLevel.authPriv,
    authProtocol: snmp.AuthProtocols['sha'],
    authKey: "authkey",
    privProtocol: snmp.PrivProtocols['aes'],
    privKey: "privkey"
};

// passing an object named userObj
exporter.createSession = async function createSession(**userObj**, configObj)

**Comment by AlexanderKormilin** [comment]
Got it, thanks.

The line

```
exporter.createSession = async function createSession(userObj, configObj)
```

is not an invocation, its just a declaration of the function responsible for the session creation.
the actual invocation happens on the line

```
const session = await this.createSession(user, config);
```

and here we can see our user object being used)

Anyway, thank you for the attention and in the most part - for the answer "it should work". It made me think there must be some error on my side.

Not sure what it was, but i dont see any issues since, i guess i should close that issue.

Thanks everyone, one final time.

**Comment by markabrahams** [comment]
Thanks for rounding this off @AlexanderKormilin - and glad you came to a working solution.

Just to confirm for reference - I've tested locally and that code you posted does indeed work fine.
